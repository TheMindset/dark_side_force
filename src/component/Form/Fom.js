import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { Link, Redirect } from 'react-router-dom'

class Form extends Component {
  constructor(getFormData) {
    super(getFormData)

    this.state = {
      name: '',
      quote: '',
      rank: '',
      nameErr: '',
      quoteErr: '',
      rankErr: false
    }
  }

  handleChange = (event) => {
    const { name, value } = event.target
    this.setState({ [name]: value })
  }

  handleSubmit = (e) => {
    e.preventDefault()

    if(!this.state.name) {
      this.setState({ nameErr: true })
    } else {
      this.setState({ nameErr: false })
    }

    if(!this.state.quote) {
      this.setState({ quoteErr: true })
    } else {
      this.setState({ quoteErr: false })
    }

    if (this.state.name && this.state.quote) {
      const user = {
        name: this.state.name,
        quote: this.state.quote,
        rank: this.state.rank
      }
      this.props.getFormData(user)
      this.setState({ isComplete: true })
    }
  }


  render() {
    const {name, quote, rank, nameErr, quoteErr, rankErr} = this.state

    let nameClass = nameErr ? 'error' : ''
    let quoteClass = quoteErr ? 'error': ''
    let rankClass = rankErr ? 'error': ''

    return (
      <form>
        <label htmlFor="name">Name</label>
        <input 
          autoFocus 
          value={name} 
          name="name" 
          placeholder="Enter a name"
          className={nameClass}
          onChange={this.handleChange}
        />
        <div>
          {nameErr && <p>No Name</p>}
        </div>

        <label htmlFor="quote">Your favorite quote</label>
        <input 
          autoFocus 
          value={quote} 
          name="quote" 
          placeholder="Enter your favorite quote"
          className={quoteClass}
          onChange={this.handleChange}
        />
        <div>
          {quoteErr && <p>No Quote</p>}
        </div>

        <label htmlFor="rank">Rank</label>
        <select 
          value={rank} 
          name="rank" 
          className={rankClass}
          onChange={this.handleChange}
        >
          <option value="Padawan">Padawan</option>
          <option value="Jedi Knight">Jedi Knight</option>
          <option value="Jedi Master">Jedi Mastern</option>
        </select>
        <div>
          {rankErr && <p>No Rank</p>}
        </div>

        <Link to='/movies' onClick={this.handleSubmit}>
          <button type="button"> May the dark force Be With You</button>
        </Link>

      </form>
    )
  }
}

export default Form