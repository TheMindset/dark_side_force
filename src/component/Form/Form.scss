form {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  height: 100vh;
  input {
    margin: 10px;
    width: 20%;
    padding: 10px;
  }

  select {
    height: 6%;
    margin: 10px;
    width: 20%;
  }

  div {
    min-height: 40px;
    p {
      display: inline;
    }
  }

  button {
    margin-top: 10px;
    &:hover {
      outline: 3px solid black;
    }
  }
}