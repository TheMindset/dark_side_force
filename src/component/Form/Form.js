import React, { Component } from 'react'
import './Form.scss'
import PropTypes from 'prop-types'
import { Link, Redirect } from 'react-router-dom'

class Form extends Component {
  constructor(getFormData) {
    super(getFormData)

    this.state = {
      name: '',
      quote: '',
      rank: 'Padawan',
      nameErr: '',
      quoteErr: '',
      isComplete: false
    }
  }

  handleChange = (event) => {
    const { name, value } = event.target
    this.setState({ [name]: value })
  }

  handleSubmit = (e) => {
    e.preventDefault()

    if(!this.state.name) {
      this.setState({ nameErr: true })
    } else {
      this.setState({ nameErr: false })
    }

    if(!this.state.quote) {
      this.setState({ quoteErr: true })
    } else {
      this.setState({ quoteErr: false })
    }

    if (this.state.name && this.state.quote) {
      const user = {
        name: this.state.name,
        quote: this.state.quote,
        rank: this.state.rank
      }
      this.props.getFormData(user)
      this.setState({ isComplete: true })
    }
  }


  render() {
    if (this.state.isComplete) {
      return <Redirect to='/movies'/>
    }
    const { name, quote, rank, nameErr, quoteErr } = this.state

    let nameClass = nameErr ? 'error' : ''
    let quoteClass = quoteErr ? 'error': ''

    return (
      <div className="login-page">
        <form className="form-background">
          <input 
            autoFocus 
            value={name} 
            name="name" 
            placeholder="Enter a name"
            className={nameClass}
            onChange={this.handleChange}
          />
          <div>
            {nameErr && <p>No Name</p>}
          </div>

          <input 
            value={quote} 
            name="quote" 
            placeholder="Enter your favorite quote"
            className={quoteClass}
            onChange={this.handleChange}
          />
          <div>
            {quoteErr && <p>No Quote</p>}
          </div>

          <select 
            value={rank} 
            name="rank" 
            onChange={this.handleChange}
          >
            <option value="Padawan">Padawan</option>
            <option value="Jedi Knight">Jedi Knight</option>
            <option value="Jedi Master">Jedi Mastern</option>
          </select>

            <button type="button" className="login-btn" onClick={this.handleSubmit}> May the dark force Be With You</button>
        </form>
      </div>
    )
  }
}

export default Form

Form.propTypes = {
  getFormData: PropTypes.func.isRequired
}