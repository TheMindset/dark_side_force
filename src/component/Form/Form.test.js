import React from 'react'
import { shallow } from 'enzyme'
import Form from './Form'

describe('Form', () => {
  let wrapper, mockGetFormData

  beforeEach(() => {
    mockGetFormData = jest.fn()
    wrapper = shallow(<Form getFormData={mockGetFormData}/>)
  })

  test('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })
  
  test('should update state when handleChange is called', () => {
    const mockEvent = { target: { name: 'name', value: 'Hola' }, preventDefault: jest.fn() }

    const expected = 'Hola'
    wrapper.instance().handleChange(mockEvent)
    expect(wrapper.state('name')).toEqual(expected)
  })

  test('should call getFormData when Submit is clicked', () => {
    const prevDefaultEvent = { preventDefault: jest.fn() }

    wrapper.setState({
      name: 'Hola',
      quote: 'I am the best',
      rank: 'Padawan'
    })

    wrapper.find('button').simulate('click', prevDefaultEvent)

    expect(mockGetFormData).toHaveBeenCalled()
  })
  
  
})
