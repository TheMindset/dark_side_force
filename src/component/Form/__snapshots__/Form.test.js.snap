// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form should match the snapshot 1`] = `
<div
  className="login-page"
>
  <form
    className="form-background"
  >
    <input
      autoFocus={true}
      className=""
      name="name"
      onChange={[Function]}
      placeholder="Enter a name"
      value=""
    />
    <div />
    <input
      className=""
      name="quote"
      onChange={[Function]}
      placeholder="Enter your favorite quote"
      value=""
    />
    <div />
    <select
      name="rank"
      onChange={[Function]}
      value="Padawan"
    >
      <option
        value="Padawan"
      >
        Padawan
      </option>
      <option
        value="Jedi Knight"
      >
        Jedi Knight
      </option>
      <option
        value="Jedi Master"
      >
        Jedi Mastern
      </option>
    </select>
    <button
      className="login-btn"
      onClick={[Function]}
      type="button"
    >
       May the dark force Be With You
    </button>
  </form>
</div>
`;
